### DISCUSSION TITLE
Running OpenDataCam in the Browser -

### DISCUSSION THREAD

**Discussion by vsaw** [opendatacam/opendatacam#533] [Ideas]
I'm currently playing with the Idea to make OpenDataCam run in the browser. For this I'm looking for volunteers that want want to run a simple benchmark on their devices to help me figure out if desktop and mobile devices provide sufficient performance.

If you want to help, visit https://nonocam.github.io/poc-tfjs-demo/ and run the benchmark!

**Comment by BarcaBear** [comment]
iPhone 12 64 GB version 15.1

Safari - 12/14fps
Chrome - 12/14fps

**Reply by vsaw** [reply]
Thank you 🙏

**Reply by BarcaBear** [reply]
Hi, I'm interested in taken up some of the development work to create a proof of concept. Where should I start. Would we host opendatacam online like AWS and load the model in the browser when let's say you goto opendatacam.com. Or would we need some sort of browser extension?

Do you have a repo we can look at with something started?

**Reply by vsaw** [reply]
Hi @BarcaBear,

awesome, thank you for offering your help!

So far there is no clear path to bring Web-Based counting to ODC. The demo at https://nonocam.github.io/poc-tfjs-demo/ was just a quick test to see if it's feasible to run object detection in the browser. First tests look promising, but I would still need to run more tests to make sure it's really feasible (e.g. long term tests to make sure phones don't throttle CPU/GPU to prevent overheating, include object tracking on top of detection, etc.)

The Demo at https://nonocam.github.io/poc-tfjs-demo/ runs is licensed under MIT, so feel free to extend it to include object tracking and run long term tests. Once we conclude that it is in fact feasible we can create a plan how to integrate this into main ODC.

As for your question, at the moment it runs in the Browser without any extension. All it needs is a Webserver to serve the page, and a Browser with Webcam access that can run [TensorFlow.js](https://www.tensorflow.org/js), which are basically all modern browsers.

Best,

**Reply by BarcaBear** [reply]
Thanks for the response!

Yes this what I'm trying to wrap my head around. We host opendatacam on a server (AWS). It loads in a web browser on a users device. Which in turn loads tensorflow.js locally using the local devices resources. 

This is all fine to this point.

Should we send the meta data to the server hosting opendatacam (AWS) or do you have an idea on how we can process the meta data locally on the users device, and push the counts to the server hosting opendatacam (AWS)

When I say meta data, I refer to the data needed for the counting logic.

I don't like the idea of pushing the meta data to the server, it should be processed locally and just push the counts to the server, ideally. 

This is why I mentioned the browser extension. Could/should the logic be there.

We will make a start on a POC and see how it goes.

Thanks!

**Reply by vsaw** [reply]
> Should we send the meta data to the server hosting opendatacam (AWS) or do you have an idea on how we can process the meta data locally on the users device, and push the counts to the server hosting opendatacam (AWS)
> 
> When I say meta data, I refer to the data needed for the counting logic.

I think you can run node-moving-things tracker on the end device as well. It should compile for the browser and it does not need a lot of compute. 

Likewise for the counting logic. This should also work locally. 

> I don't like the idea of pushing the meta data to the server, it should be processed locally and just push the counts to the server, ideally.

I agree. I think it is best just to push/update the counter state (E.g. to a special REST endpoint on the server). This should be much better for the bandwidth then to push raw detections or tracker data.  

> This is why I mentioned the browser extension. Could/should the logic be there.

I don’t think you need a browser extension. All of this should be able to run in the browser without any extension. 

> We will make a start on a POC and see how it goes.

Great. Keep me posted :)

**Comment by munsterlander** [comment]
Android - Samsung Galaxy S21
Chrome 14-18 fps

**Reply by vsaw** [reply]
Thank you 🙏. Already added to the list https://github.com/nonocam/poc-tfjs-demo

**Comment by taz4535** [comment]
Chrome: 30 fps

**Reply by vsaw** [reply]
Thanks, Which device are you running?

**Comment by taz4535** [comment]
ah my bad. missed the specific requirement of mobile device.
